docs/architecture/grpc-for-wcf-developers/migrating-request-reply.md

```diff
@@ -0,0 +1,418 @@
+}
+
+service Portfolios {
+    rpc Get(GetRequest) returns (GetResponse);
+}
+```
+
+The WCF `GetAll` method only takes a single parameter, `traderId`, so it might seem that one could specify `string` as the parameter type, but gRPC requires a defined message type. This requirement helps to enforce the practice of using custom messages for all inputs and outputs, for future backward compatibility.
+
+The WCF method also returned a `List<Portfolio>`, but for the same reason it doesn't allow simple parameter types, gRPC won't allow `repeated Portfolio` as a return type. Instead, create a `GetAllResponse` type to wrap the list.
+
+> [!WARNING]
+> You may be tempted to create a `PortfolioList` message or similar and use it across multiple service methods, but you should resist this temptation. It's impossible to know how the various methods on a service may evolve in the future, so keep their messages specific and cleanly separated.
+
+```protobuf
+message GetAllRequest {
+    string traderId = 1;
+}
+
+message PortfolioList {
+    repeated Portfolio portfolios = 1;
+}
+
+service Portfolios {
+    rpc Get(GetRequest) returns (Portfolio);
+    rpc GetAll(GetAllRequest) returns (GetAllResponse);
+}
+```
+
+If you save your project with these changes, the gRPC build target will run in the background and generate all the Protobuf message types, and a base class you can inherit to implement the service.
+
+Open up the `Services/GreeterService.cs` class and delete the example code. Now you can add the Portfolio service implementation. The generated base class will be in the `Protos` namespace, and is actually generated as a nested class. gRPC creates a static class with the same name as the service in the `.proto` file, and then a base class with the suffix `Base` inside that static class, so the full identifier for the base type is `TraderSys.Portfolios.Protos.Portfolios.PortfoliosBase`.
+
+```csharp
+namespace TraderSys.Portfolios.Services
+{
+    public class PortfolioService : Protos.Portfolios.PortfoliosBase
+    {
+    }
+}
+```
+
+The base class declares `virtual` methods for `Get` and `GetAll` that can be overridden to implement the service. The methods are `virtual` rather than `abstract` so that if you don't implement them, the service can return an explicit gRPC `Unimplemented` status code, much like you might throw a `NotImplementedException` in regular C# code.
+
+The signature for all gRPC service methods in ASP.NET Core is consistent. There are two parameters: the first is the message type declared in the `.proto` file, and the second is a `ServerCallContext` that works similarly to the `HttpContext` from ASP.NET Core. (In fact, there is an extension method `GetHttpContext` on the `ServerCallContext` type that you can use to get the underlying `HttpContext`, although you shouldn't need to use it very often.) There will be a look at `ServerCallContext` later in the chapter, and also in the chapter on authentication.
```

Comment:
The signature is consitent for all unary methods. Signatures vary between method types.